array.go

```diff
@@ -913,3 +913,40 @@ func countElement(array []interface{}, element interface{}) int {
 	}
 	return count
 }
+
+func (a *Array) Filter(filter func(index int, value *Value) bool) *Array {
+	a.chain.enter("Filter()")
+	defer a.chain.leave()
+
+	if a.chain.failed() {
+		return a
+	}
+
+	if filter == nil {
+		a.chain.fail(AssertionFailure{
+			Type: AssertUsage,
+			Errors: []error{
+				errors.New("unexpected nil function argument"),
+			},
+		})
+	}
+
+	var filteredArray []interface{}
+
+	chainFailure := false
+	for index, element := range a.value {
+		valueChain := a.chain.clone()
+		valueChain.setFailCallback(func() {
+			chainFailure = true
+		})
+		if filter(index, newValue(valueChain, element)) {
+			filteredArray = append(filteredArray, element)
+		}
+	}
+
+	if chainFailure {
+		a.chain.setFailed()
+	}
```

Comment:
In Filter(), in contrast to Every(), if the predicate fails, we should exclude the element from array, but we should not mark the array as failed. And we should use setFatal(false) to make failures inside the predicate non-fatal. See code snippets in #169.

And we'll to cover this behavior with tests.

Reply:
So if I understand correctly, we don't need to use the setFailCallback function, is that right?

Reply:
Nope, we still need setFailCallback to detect that some assertion inside predicate failed. We only need to change our handling of this case: instead of marking original array as failed, we should just skip failed element.

Reply:
But in this case if we don't use the setFailed method, under what condition can the chain fail and what purpose does chainFailure serve?

Reply:
Quote from the issue:

> Predicate is considered to succeed if BOTH conditions are met:
>
>    * assertions inside predicate did not fail (like value.String().NotEmpty() in the above example)
>    * predicate returned true

Imagine you have this call:

```go
object.Filter(func(key string, value *httpexpect.Value) bool {
    // make assertions
    value.String().NotEmpty()

    // return boolean
    return true
})
```

What Filter() should do for every value:

* check if any assertion failed **inside** predicate, in this case `value.String().NotEmpty()`
* check if predicate return true
* if at least one of the above is false, Filter() should skip this element

Important thing: if an assertion failed **inside predicate**, like `value.String().NotEmpty()` in this example, it does not mean that the test should fail. In this context failure means only "skip this element".

To achieve this, we need to do a few things:

* we need to clone the chain and pass this clone to predicate
* we need to setup the cloned chain as non-fatal (call setFalal(false)), so that assertions made in predicate would not cause failure of the whole test
* we need to be informed that the chain, or one of its children, failed inside the predicate; this can be done using setFailureCallback; if an assertion on value will fail inside predicate, this callback will be invoked and we will mark predicated as failed

No matter of predicate results, we should NOT call `a.chain.setFailed()` (where `a` is the original array). This will mark the whole array as failed and cause the whole test to fail, which we don't want.